In [ ]:
import os
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.decomposition import PCA

sns.set(style="whitegrid")

## Đọc dữ liệu

In [ ]:
print("Đọc dữ liệu từ file CSV...")
df = pd.read_csv("./datasets/preprocessing_data/clean_data.csv")

Đọc dữ liệu từ file CSV...


<ipython-input-6-6bbabb295e53>:2: DtypeWarning: Columns (2050) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv("./datasets/preprocessing_data/clean_data.csv")


## Tổng quan dữ liệu
- Xem 5 dòng đầu  tiên của dữ liệu
- Thông tin tổng quan về kiểu dữ liệu và số lượng giá trị không null
- Thống kê mô tả các cột số
- Kiểu dữ liệu từng cột

In [ ]:
print("1. Kích thước dữ liệu:", df.shape)
print("\n2. 5 dòng đầu:")
display(df.head())

print("\n3. Thông tin dữ liệu:")
df.info()

print("\n4. Thống kê mô tả:")
display(df.describe(include='all'))

print("\n5. Kiểu dữ liệu từng cột:")
print(df.dtypes)

1. Kích thước dữ liệu: (73265, 2055)

2. 5 dòng đầu:


f0        f1        f2        f3        f4        f5        f6  \
0  0.162773  0.007738  0.016945  1.544321  0.017166  0.167690  0.217837   
1  0.374471  0.843741  3.395411  2.154582  0.147768  2.960643  0.382921   
2  0.000000  0.255170  0.666132  0.250708  0.000000  0.000000  0.421206   
3  0.000000  1.376298  0.044668  2.194536  0.000000  0.073896  0.146056   
4  0.896031  0.173879  0.338651  0.543566  0.000000  1.051726  0.850426   

         f7        f8        f9  ...     f2045     f2046     f2047  label  \
0  0.084524  0.442217  0.261509  ...  0.042974  0.314448  0.020068     19   
1  0.018443  0.006840  1.031437  ...  0.030779  0.074114  0.632746     34   
2  0.334652  0.149745  0.000000  ...  1.118111  0.014624  0.000000      2   
3  0.000000  0.647502  0.280085  ...  0.023793  0.221037  0.013904     17   
4  1.307167  0.204651  0.647726  ...  0.185884  0.110241  0.015055     18   

               split  filename      class_name  class_fullname       PCA1  \
0  train_oversampled       NaN         gorilla         gorilla   2.088935   
1  train_oversampled       NaN          weasel          weasel   3.186141   
2  train_oversampled       NaN    killer whale    killer whale  18.082210   
3  train_oversampled       NaN  humpback whale  humpback whale  14.242381   
4  train_oversampled       NaN        elephant        elephant  -5.060207   

        PCA2  
0  -1.376181  
1  -8.506805  
2  15.854512  
3  13.289589  
4   6.940943  

[5 rows x 2055 columns]


3. Thông tin dữ liệu:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 73265 entries, 0 to 73264
Columns: 2055 entries, f0 to PCA2
dtypes: float64(2050), int64(1), object(4)
memory usage: 1.1+ GB

4. Thống kê mô tả:


f0            f1            f2            f3            f4  \
count   73265.000000  73265.000000  73265.000000  73265.000000  73265.000000   
unique           NaN           NaN           NaN           NaN           NaN   
top              NaN           NaN           NaN           NaN           NaN   
freq             NaN           NaN           NaN           NaN           NaN   
mean        0.354701      0.658329      0.657843      0.798829      0.358392   
std         0.531786      0.731448      0.933524      1.090671      0.599066   
min         0.000000      0.000000      0.000000      0.000000      0.000000   
25%         0.023506      0.131985      0.049549      0.089524      0.014381   
50%         0.145575      0.425976      0.296255      0.375915      0.115609   
75%         0.469114      0.941413      0.881519      1.062402      0.420903   
max         6.040072      9.615986      9.656792     12.311359      7.639313   

                  f5            f6            f7            f8            f9  \
count   73265.000000  73265.000000  73265.000000  73265.000000  73265.000000   
unique           NaN           NaN           NaN           NaN           NaN   
top              NaN           NaN           NaN           NaN           NaN   
freq             NaN           NaN           NaN           NaN           NaN   
mean        0.796312      0.648085      0.455565      0.374246      0.496781   
std         1.133570      0.741044      0.836797      0.569641      0.809943   
min         0.000000      0.000000      0.000000      0.000000      0.000000   
25%         0.064289      0.109698      0.018492      0.037951      0.051911   
50%         0.341315      0.395325      0.143366      0.163694      0.213026   
75%         1.067045      0.937120      0.524272      0.463244      0.583743   
max        12.931930      7.721712     12.733008      7.141273     14.015243   

        ...         f2045         f2046         f2047         label  \
count   ...  73265.000000  73265.000000  73265.000000  73265.000000   
unique  ...           NaN           NaN           NaN           NaN   
top     ...           NaN           NaN           NaN           NaN   
freq    ...           NaN           NaN           NaN           NaN   
mean    ...      0.757847      0.375826      0.306121     24.599823   
std     ...      0.976601      0.706920      0.454534     14.417264   
min     ...      0.000000      0.000000      0.000000      0.000000   
25%     ...      0.073930      0.027250      0.023523     12.000000   
50%     ...      0.373856      0.132875      0.136431     25.000000   
75%     ...      1.082436      0.399749      0.394944     37.000000   
max     ...      9.146830     10.189293      5.034893     49.000000   

                    split                filename  class_name  class_fullname  \
count               73265                    7465       73265           73265   
unique                  2                    7465          50              50   
top     train_oversampled  killer+whale_10132.jpg       horse           horse   
freq                65800                       1        1645            1645   
mean                  NaN                     NaN         NaN             NaN   
std                   NaN                     NaN         NaN             NaN   
min                   NaN                     NaN         NaN             NaN   
25%                   NaN                     NaN         NaN             NaN   
50%                   NaN                     NaN         NaN             NaN   
75%                   NaN                     NaN         NaN             NaN   
max                   NaN                     NaN         NaN             NaN   

                PCA1          PCA2  
count   7.326500e+04  7.326500e+04  
unique           NaN           NaN  
top              NaN           NaN  
freq             NaN           NaN  
mean    1.241377e-16 -2.482754e-16  
std     7.735311e+00  7.276225e+00  
min    -1.9


5. Kiểu dữ liệu từng cột:
f0                float64
f1                float64
f2                float64
f3                float64
f4                float64
                   ...   
filename           object
class_name         object
class_fullname     object
PCA1              float64
PCA2              float64
Length: 2055, dtype: object


## Kiểm tra số lượng giá trị thiếu và các dòng bị trùng lặp

In [ ]:
print("\n6. Số lượng giá trị thiếu:")
missing = df.isnull().sum()
print(missing[missing > 0])

print("\n7. Dòng bị trùng:", df.duplicated().sum())


6. Số lượng giá trị thiếu:
filename    65800
dtype: int64

7. Dòng bị trùng: 35937


In [ ]:
!python3 -m pip -V

pip 24.1.2 from /usr/local/lib/python3.11/dist-packages/pip (python 3.11)


In [ ]:
!python3 -m pip install plotly seaborn umap-learn

In [ ]:

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px
import plotly.graph_objects as go
from plotly.subplots import make_subplots
from sklearn.decomposition import PCA
from sklearn.manifold import TSNE
import umap
from sklearn.preprocessing import StandardScaler
import warnings
warnings.filterwarnings('ignore')

# Thiết lập style cho matplotlib
plt.style.use('default')
sns.set_palette("husl")

In [ ]:


class DataVisualizationTool:
    def __init__(self, df):
        """
        Khởi tạo công cụ trực quan hóa
        df: DataFrame chứa dữ liệu
        """
        self.df = df.copy()
        self.feature_cols = [col for col in df.columns if col.startswith('f') and col != 'filename']
        self.X = df[self.feature_cols].values
        self.y = df['class_name'].values
        self.labels = df['label'].values

        # Chuẩn hóa dữ liệu
        #self.scaler = StandardScaler()
        #self.X_scaled = self.scaler.fit_transform(self.X)
#Sửa lại
        self.X_scaled = self.X

        print(f"Số lượng features: {len(self.feature_cols)}")
        print(f"Số lượng samples: {len(self.df)}")
        print(f"Số lượng classes: {len(self.df['class_name'].unique())}")

    def plot_pca_analysis(self, n_components=10):
        """
        Phân tích PCA và vẽ biểu đồ explained variance
        """
        pca = PCA(n_components=n_components)
        pca.fit(self.X_scaled)

        # Tạo subplot
        fig, axes = plt.subplots(2, 2, figsize=(15, 12))
        fig.suptitle('Phân tích PCA', fontsize=16, fontweight='bold')

        # 1. Explained Variance Ratio
        axes[0,0].bar(range(1, n_components+1), pca.explained_variance_ratio_)
        axes[0,0].set_title('Tỉ lệ Phương sai Giải thích của từng PC')
        axes[0,0].set_xlabel('Principal Component')
        axes[0,0].set_ylabel('Explained Variance Ratio')

        # 2. Cumulative Explained Variance
        cumsum = np.cumsum(pca.explained_variance_ratio_)
        axes[0,1].plot(range(1, n_components+1), cumsum, 'bo-')
        axes[0,1].set_title('Phương sai Giải thích Tích lũy')
        axes[0,1].set_xlabel('Principal Component')
        axes[0,1].set_ylabel('Cumulative Explained Variance')
        axes[0,1].grid(True, alpha=0.3)

        # 3. PCA transformation và scatter plot
        X_pca = pca.transform(self.X_scaled)

        # Nếu đã có PCA1, PCA2 trong data thì sử dụng
        if 'PCA1' in self.df.columns and 'PCA2' in self.df.columns:
            scatter = axes[1,0].scatter(self.df['PCA1'], self.df['PCA2'],
                                      c=self.labels, cmap='tab20', alpha=0.7)
            axes[1,0].set_xlabel('PCA1')
            axes[1,0].set_ylabel('PCA2')
        else:
            scatter = axes[1,0].scatter(X_pca[:, 0], X_pca[:, 1],
                                      c=self.labels, cmap='tab20', alpha=0.7)
            axes[1,0].set_xlabel('PC1')
            axes[1,0].set_ylabel('PC2')

        axes[1,0].set_title('PCA Scatter Plot (2D)')
        plt.colorbar(scatter, ax=axes[1,0])

        # 4. Feature importance của PC1 và PC2
        feature_importance = np.abs(pca.components_[:2]).mean(axis=0)
        top_features_idx = np.argsort(feature_importance)[-20:]  # Top 20 features

        axes[1,1].barh(range(len(top_features_idx)), feature_importance[top_features_idx])
        axes[1,1].set_yticks(range(len(top_features_idx)))
        axes[1,1].set_yticklabels([f'f{i}' for i in top_features_idx])
        axes[1,1].set_title('Top 20 Features quan trọng nhất (PC1 & PC2)')
        axes[1,1].set_xlabel('Importance Score')

        plt.tight_layout()
        plt.show()

        return pca, X_pca

    def plot_interactive_scatter(self, method='pca', n_components=2):
        """
        Tạo scatter plot tương tác với Plotly
        method: 'pca', 'tsne', 'umap'
        """
        if method == 'pca':
            if 'PCA1' in self.df.columns and 'PCA2' in self.df.columns:
                # Sử dụng PCA có sẵn
                X_reduced = self.df[['PCA1', 'PCA2']].values
                title = "PCA Visualization (Có sẵn)"
                x_label, y_label = "PCA1", "PCA2"
            else:
                # Tính PCA mới
                pca = PCA(n_components=n_components)
                X_reduced = pca.fit_transform(self.X_scaled)
                title = f"PCA Visualization (Tính toán mới)"
                x_label, y_label = "PC1", "PC2"

        elif method == 'tsne':
            print("Đang chạy t-SNE... (có thể mất vài phút)")
            tsne = TSNE(n_components=2, random_state=42, perplexity=30)
            X_reduced = tsne.fit_transform(self.X_scaled)
            title = "t-SNE Visualization"
            x_label, y_label = "t-SNE 1", "t-SNE 2"

        elif method == 'umap':
            print("Đang chạy UMAP...")
            reducer = umap.UMAP(n_components=2, random_state=42)
            X_reduced = reducer.fit_transform(self.X_scaled)
            title = "UMAP Visualization"
            x_label, y_label = "UMAP 1", "UMAP 2"

        # Tạo DataFrame cho Plotly
        plot_df = pd.DataFrame({
            'x': X_reduced[:, 0],
            'y': X_reduced[:, 1],
            'class': self.y,
            'label': self.labels,
            'split': self.df['split'] if 'split' in self.df.columns else 'unknown'
        })

        # Tạo interactive plot
        fig = px.scatter(plot_df, x='x', y='y', color='class',
                        hover_data=['label', 'split'],
                        title=title,
                        labels={'x': x_label, 'y': y_label})

        fig.update_layout(
            width=900, height=700,
            title_font_size=16,
            showlegend=True
        )

        fig.show()

        return X_reduced

    def plot_class_distribution(self):
        """
        Vẽ phân phối các class
        """
        fig, axes = plt.subplots(1, 2, figsize=(15, 6))

        # 1. Class distribution
        class_counts = self.df['class_name'].value_counts()
        axes[0].bar(range(len(class_counts)), class_counts.values)
        axes[0].set_xticks(range(len(class_counts)))
        axes[0].set_xticklabels(class_counts.index, rotation=45, ha='right')
        axes[0].set_title('Phân phối số lượng mẫu theo Class')
        axes[0].set_ylabel('Số lượng mẫu')

        # 2. Train/Test split nếu có
        if 'split' in self.df.columns:
            split_counts = self.df['split'].value_counts()
            colors = ['#FF6B6B', '#4ECDC4', '#45B7D1', '#96CEB4']
            axes[1].pie(split_counts.values, labels=split_counts.index,
                       autopct='%1.1f%%', colors=colors[:len(split_counts)])
            axes[1].set_title('Phân chia Train/Test')
        else:
            axes[1].text(0.5, 0.5, 'Không có thông tin\nphân chia dữ liệu',
                        ha='center', va='center', transform=axes[1].transAxes)
            axes[1].set_title('Train/Test Split')

        plt.tight_layout()
        plt.show()

    def plot_feature_analysis(self, sample_features=50):
        """
        Phân tích một số features ngẫu nhiên
        """
        # Chọn ngẫu nhiên một số features
        selected_features = np.random.choice(self.feature_cols,
                                           min(sample_features, len(self.feature_cols)),
                                           replace=False)

        fig, axes = plt.subplots(2, 2, figsize=(15, 12))
        fig.suptitle(f'Phân tích {len(selected_features)} Features ngẫu nhiên',
                     fontsize=16, fontweight='bold')

        # 1. Distribution của features
        feature_data = self.df[selected_features].values.flatten()
        axes[0,0].hist(feature_data, bins=50, alpha=0.7, color='skyblue')
        axes[0,0].set_title('Phân phối giá trị của các Features')
        axes[0,0].set_xlabel('Giá trị Feature')
        axes[0,0].set_ylabel('Tần suất')

        # 2. Correlation heatmap của một số features
        sample_size = min(20, len(selected_features))
        sample_features_small = selected_features[:sample_size]
        corr_matrix = self.df[sample_features_small].corr()

        sns.heatmap(corr_matrix, annot=False, cmap='coolwarm', center=0, ax=axes[0,1])
        axes[0,1].set_title(f'Ma trận tương quan ({sample_size} features)')

        # 3. Feature variance
        feature_vars = self.df[selected_features].var().sort_values(ascending=False)
        axes[1,0].bar(range(len(feature_vars)), feature_vars.values)
        axes[1,0].set_title('Phương sai của các Features')
        axes[1,0].set_xlabel('Feature Index')
        axes[1,0].set_ylabel('Variance')

        # 4. Box plot của một số features theo class
        sample_classes = self.df['class_name'].unique()[:5]  # Lấy 5 class đầu
        sample_features_box = selected_features[:5]  # Lấy 5 features đầu

        box_data = []
        for feature in sample_features_box:
            for class_name in sample_classes:
                class_data = self.df[self.df['class_name'] == class_name][feature].values
                for value in class_data:
                    box_data.append({'feature': feature, 'class': class_name, 'value': value})

        box_df = pd.DataFrame(box_data)
        if len(box_df) > 0:
            sns.boxplot(data=box_df, x='feature', y='value', hue='class', ax=axes[1,1])
            axes[1,1].set_title('Box Plot: Features theo Class')
            axes[1,1].tick_params(axis='x', rotation=45)

        plt.tight_layout()
        plt.show()

    def create_comprehensive_dashboard(self):
        """
        Tạo dashboard tổng quan
        """
        print("=== DASHBOARD TRỰC QUAN HÓA DỮ LIỆU ĐA CHIỀU ===\n")

        # 1. Thống kê cơ bản
        print("1. THỐNG KÊ CƠ BẢN:")
        print(f"   - Số features: {len(self.feature_cols)}")
        print(f"   - Số samples: {len(self.df)}")
        print(f"   - Số classes: {len(self.df['class_name'].unique())}")
        print(f"   - Classes: {', '.join(self.df['class_name'].unique())}")

        # 2. Phân phối class
        print("\n2. PHÂN PHỐI CLASSES:")
        self.plot_class_distribution()

        # 3. PCA Analysis
        print("\n3. PHÂN TÍCH PCA:")
        pca, X_pca = self.plot_pca_analysis()

        # 4. Interactive visualization
        print("\n4. TRỰC QUAN HÓA TƯƠNG TÁC:")

        # PCA
        print("   Đang tạo PCA plot...")
        self.plot_interactive_scatter(method='pca')

        # t-SNE
        print("   Đang tạo t-SNE plot...")
        self.plot_interactive_scatter(method='tsne')

        # UMAP
        print("   Đang tạo UMAP plot...")
        self.plot_interactive_scatter(method='umap')

        # 5. Feature analysis
        print("\n5. PHÂN TÍCH FEATURES:")
        self.plot_feature_analysis()

        print("\n=== HOÀN THÀNH DASHBOARD ===")


In [ ]:

viz_tool = DataVisualizationTool(df)
viz_tool.create_comprehensive_dashboard()

Output hidden; open in https://colab.research.google.com to view.

## 1. Thống Kê Cơ Bản

- **Số lượng đặc trưng (features)**: 2,048 chiều trích xuất từ mô hình ResNet101.
- **Số lượng mẫu (samples)**: ~66,000 ảnh động vật sau cân bằng lớp.
- **Số lớp (classes)**: 50 loài động vật khác nhau.

-> *Dữ liệu có quy mô lớn, số chiều cao và đa dạng lớp. Do đó, cần thực hiện chuẩn hóa và giảm chiều để tăng hiệu quả huấn luyện và trực quan hóa.*

---

## 2. Phân Phối Class và Tập Dữ Liệu

### **Biểu đồ 1: Phân phối số lượng mẫu theo class**

- Tất cả các lớp được phân phối **tương đối đồng đều**.
- Không xuất hiện hiện tượng mất cân bằng nghiêm trọng.

-> *Kỹ thuật cân bằng lớp đã đảm bảo mô hình không bị thiên lệch theo các lớp nhiều mẫu.*

---

### **Biểu đồ 2: Phân chia Train/Test**

- Tỷ lệ phân chia: `~90% train_oversampled` và `~10% test`.
- Tập train được cân bằng theo lớp, còn test giữ nguyên phân phối ban đầu.

-> *Chiến lược này giúp đảm bảo tổng quát hóa và công bằng trong đánh giá mô hình.*

---

## 3. Phân Tích PCA (Giảm Chiều)

### **Biểu đồ 3: Tỉ lệ phương sai giải thích của từng thành phần**

- PC1 giữ ~5.8% phương sai, 10 PCs đầu tiên giữ ~30%.
- Thông tin được phân tán đều giữa nhiều thành phần.

-> *PCA không phù hợp để giảm xuống quá ít chiều → chỉ nên dùng cho trực quan hóa.*

---

### **Biểu đồ 4: Phương sai tích lũy**

- 10 thành phần đầu chỉ đạt ~30% tổng thông tin.
- Cần giữ nhiều thành phần hơn nếu dùng PCA trong pipeline mô hình hóa.

---

### **Biểu đồ 5: PCA Scatter Plot (2D)**

- Các điểm dữ liệu giữa các lớp bị **chồng lấn đáng kể**.
- Không có sự tách biệt rõ ràng trên không gian tuyến tính PCA1–PCA2.

---

### **Biểu đồ 6: Top 20 Feature quan trọng theo PC1 & PC2**

- Một số feature như `f217`, `f504`, `f1294` có ảnh hưởng lớn nhất.

-> *Có thể sử dụng các feature quan trọng này trong các mô hình nhẹ hoặc phân tích giải thích.*

---

## 4. Trực Quan Hóa Tương Tác

### **Biểu đồ 7: PCA Interactive Plot**

- Các lớp không được phân tách rõ ràng → PCA **không phù hợp để phân loại trực tiếp**.

---

### **Biểu đồ 8: t-SNE Interactive Visualization**

- Các lớp được phân cụm rõ ràng hơn, phản ánh cấu trúc phi tuyến trong dữ liệu.

-> *t-SNE phù hợp cho trực quan hóa lớp và phát hiện outlier.*

---

### **Biểu đồ 9: UMAP Interactive Visualization**

- Các cụm lớp được phân tách rõ ràng, nhiều lớp hiển thị ranh giới tách biệt.

-> *UMAP cung cấp trực quan hóa tốt nhất → có thể dùng cho giảm chiều đầu vào mô hình hoặc phân cụm.*

---

## 5. Phân Tích Features (Đặc Trưng)

### **Biểu đồ 10: Phân phối giá trị features**

- Mặc dù đã chuẩn hóa bằng PowerTransformer, nhiều feature vẫn có xu hướng **lệch phải nhẹ**, phần lớn giá trị gần 0.

---

### **Biểu đồ 11: Ma trận tương quan**

- Các feature hầu như không tương quan cao với nhau.

-> *Không có chiều trùng lặp → giữ lại toàn bộ features vẫn hợp lý.*

---

### **Biểu đồ 12: Phương sai của 50 feature ngẫu nhiên**

- Phương sai dao động trong khoảng [0.1–1.3], thể hiện mức đóng góp khác nhau.

-> *Có thể chọn lọc feature nếu cần giảm độ phức tạp.*

---

### **Biểu đồ 13: Boxplot theo class**

- Một số feature như `f1741`, `f603`, `f1294` thể hiện khả năng **phân biệt nhẹ** giữa các lớp như *elephant*, *killer whale*, *gorilla*.

-> *Các feature này có thể sử dụng trong mô hình tuyến tính hoặc kỹ thuật giải thích mô hình như SHAP.*